VSR: increase message pool capacity

sentientwaffle opened the pull request:
Fixes #69.

## Bug

Each replica `pipeline` stores `config.pipelining_max == config.clients_max` prepare messages. Each of those prepares has up to `config.replicas_max` `prepare_ok` messages.

If all messages in the pool are allocated to `prepare_ok`s of future prepares (see https://gist.github.com/sentientwaffle/dfc2dfa666a3f2e6afd3a3a16bea1df9 for an example `pipeline` state), the `prepare_ok` messages for the next-in-line (blocking) prepare are just dropped. The other replicas retry sending them, but this happens indefinitely, since the blocked replica is deadlocked.

```
[debug] (vsr): 0: prepare_timeout fired
[debug] (vsr): 0: prepare_timeout backing off
[debug] (vsr): 0: prepare_timeout after=535..949 (rtt=30 min=10 max=1000 attempts=236)
[debug] (replica): 0: on_prepare_timeout: waiting for replica 1
[debug] (replica): 0: on_prepare_timeout: waiting for replica 2
[debug] (replica): 0: on_prepare_timeout: waiting for replica 3
[debug] (replica): 0: on_prepare_timeout: waiting for replica 4
[debug] (replica): 0: on_prepare_timeout: waiting for replica 5
[debug] (replica): 0: on_prepare_timeout: replicating to replica 2
```

In the VOPR, the drop triggers the `deliver_message: target message bus has no free messages, dropping` log message.

## Fix

`config.message_bus_messages_max` must be greater than `pipelining_max × (1 + replicas_max)` (`+1` to account for the prepares, not just the `prepare_ok`s). And that doesn't include the messages that may be allocated elsewhere, e.g. I/O, quorum messages, etc.

## Seeds

Debugged using the seed:
- `14308390800935580042`: This is the one I used for debugging. Replica 0 is stuck waiting for a quorum of `prepare_ok` messages, but all of its `Message`s are already allocated in the `pipeline`.

This patch also fixes the following seeds:
- `10230879186405206692`
- `16623794497874437158`

## Misc

- Includes @jorangreef's suggested fix: make the control flow more obvious (in `client.zig`) to avoid a message leak when the `request_queue` is full.
- Also includes fix for some typos.

sentientwaffle committed:
VSR: fix message leak in client

Per Joran's suggestion, make the control flow more obvious to avoid a message leak when the `request_queue` is full.

sentientwaffle committed:
Config: avoid deadlock from too few messages

Fixes https://github.com/coilhq/tigerbeetle/issues/69.

*Bug*

Each replica `pipeline` stores `config.pipelining_max == config.clients_max` prepare messages. Each of those prepares has up to `config.replicas_max` `prepare_ok` messages.

If all messages in the pool are allocated to `prepare_ok`s of future prepares (see https://gist.github.com/sentientwaffle/dfc2dfa666a3f2e6afd3a3a16bea1df9 for an example `pipeline` state), the `prepare_ok` messages for the next-in-line (blocking) prepare are just dropped. The other replicas retry sending them, but this happens indefinitely, since the blocked replica is deadlocked.

```
[debug] (vsr): 0: prepare_timeout fired
[debug] (vsr): 0: prepare_timeout backing off
[debug] (vsr): 0: prepare_timeout after=535..949 (rtt=30 min=10 max=1000 attempts=236)
[debug] (replica): 0: on_prepare_timeout: waiting for replica 1
[debug] (replica): 0: on_prepare_timeout: waiting for replica 2
[debug] (replica): 0: on_prepare_timeout: waiting for replica 3
[debug] (replica): 0: on_prepare_timeout: waiting for replica 4
[debug] (replica): 0: on_prepare_timeout: waiting for replica 5
[debug] (replica): 0: on_prepare_timeout: replicating to replica 2
```

In the VOPR, the drop triggers the `deliver_message: target message bus has no free messages, dropping` log message.

*Fix*

`message_bus_messages_max` must be greater than `pipelining_max × (1 + replicas_max)` (`+1` to account for the prepares, not just the `prepare_ok`s). And that doesn't include the messages that may be allocated elsewhere, e.g. I/O, quorum messages, etc.

*Seeds*

Debugged using the seed:
- 14308390800935580042: This is the one I used for debugging. Replica 0 is stuck waiting for a quorum of `prepare_ok` messages, but all of its `Message`s are already allocated in the `pipeline`.
This patch also fixes the following seeds:
- 10230879186405206692
- 16623794497874437158

sentientwaffle committed:
Docs: fix typos

jorangreef commented on src/vsr/client.zig:
```diff
@@ -188,17 +188,19 @@ pub fn Client(comptime StateMachine: type, comptime MessageBus: type) type {
+
             const was_empty = self.request_queue.empty();
 
             self.request_queue.push(.{
```
Let's add `push_assume_capacity` to `RingBuffer` and then use it here.

jorangreef commented on src/config.zig:
```diff
@@ -177,6 +171,22 @@ pub const io_depth_read = 8;
 pub const io_depth_write = 8;
 
+// There must be enough messages to ensure that the replica can always progress,
+// i.e. it never deadlocks because it doesn't have a free message.
```
```suggestion
// There must be enough messages to ensure that the replica can always progress, to avoid deadlock.
```

jorangreef commented on src/config.zig:
```diff
@@ -177,6 +171,22 @@ pub const io_depth_read = 8;
+
+/// The number of full-sized messages allocated at initialization by the message bus.
+pub const message_bus_messages_max = pipelining_messages_max + quorum_messages_max +
+    loopback_queue_messages_max + client_table_messages_max + journal_messages_max;
```
Great compendium!

I think we still need to account for outgoing messages to other replicas in the send queue?

Let's also define these constants in their own block scope to avoid introducing them in the global scope here (even as private), and then we break out the value we want to assign to `message_bus_messages_max`.

In terms of assertions, we could check that the result is greater than the number of replicas at least.

Let's also explain the magic number for `3` with a comment listing the names of all the different quorums.

To make sure we're covering all possible concurrent uses of messages, we could grep the source for all `.ref()` calls to see where we store message pointers (if you haven't already, it looks like you have!)?

sentientwaffle replied:
> I think we still need to account for outgoing messages to other replicas in the send queue?

Good catch, I forgot about that and the `Connection`'s `recv_message`.

> In terms of assertions, we could check that the result is greater than the number of replicas at least.

Is that misleading? Doesn't it (incorrectly) imply that values greater than `replicas_max` are safe/ok?

Maybe derived values such as `message_bus_messages_max` don't belong in `config.zig` (i.e. move to a `constants.zig` instead). The value that is currently `message_bus_messages_max` would be defined in terms of a combination other constants (e.g. `quorum_messages_max`) and configuration values (e.g. `replicas_max`).

In other words, `config.zig` should be for values that are actually configurable.

> To make sure we're covering all possible concurrent uses of messages, we could grep the source for all .ref() calls to see where we store message pointers (if you haven't already, it looks like you have!)?

I did, but evidently missed a couple so I'll double-check.

jorangreef replied:
> Is that misleading? Doesn't it (incorrectly) imply that values greater than replicas_max are safe/ok?

It's more like the assertion is saying we expect it to be at least `replicas_max`, i.e. that's as far as we can assert. In other words, we definitely wouldn't expect it to be less than the number of replicas. Alternatively, we could just assert that's greater than zero and that we didn't get a "anything times zero is zero" bug.

> In other words, config.zig should be for values that are actually configurable.

Having `constants.zig` is a good idea.

I'm thinking eventually we might have three configs: one runtime config, one cluster init time config, and one comptime config, with some runtime configs allowed to update the corresponding comptime default. We can sketch different ways of achieving this, and reorganize this as we expand our CLI options and once we have our init config being persisted in the superblock.

jorangreef replied:
> Maybe derived values such as message_bus_messages_max don't belong in config.zig (i.e. move to a constants.zig instead). The value that is currently message_bus_messages_max would be defined in terms of a combination other constants (e.g. quorum_messages_max) and configuration values (e.g. replicas_max).

That's exactly it. It is derived and should be derived closer to where it's used.

Let's move the derived value into `message_bus.zig` or even better to the message pool. We've also been learning this for the LSM namespace, keeping the derived constants in there.

jorangreef commented on src/config.zig:
```diff
@@ -177,6 +171,22 @@ pub const io_depth_read = 8;
+
+/// The number of header-sized messages allocated at initialization by the message bus.
+/// These are much smaller/cheaper and we can therefore have many of them.
+pub const message_bus_headers_max = connections_max * connection_send_queue_max * 2;
```
Could we tighten this a little more by looking for all the places where we use header-only messages?

At present, the value I tuned here is more of a thumbsuck. It's very nice now that we're starting to work out the exact allocation we need.

jorangreef reviewed (changes requested):
Thanks @sentientwaffle this is going to be great when it lands.

jorangreef commented:
Are there any other places in the code where we `ref()` when pushing, forgetting to `unref` if the queue is full?

jorangreef commented:
Thinking some more about this — if we tackle #70 here too, then we can tighten the bounds on our `message_bus_messages_max` even further since we know then that `prepare_ok` messages will be accounted for in `message_bus_headers_max`. This will let us reduce static memory allocation considerably since the pipeline can be long.

sentientwaffle commented:
> Thinking some more about this — if we tackle #70 here too, then we can tighten the bounds on our `message_bus_messages_max` even further since we know then that `prepare_ok` messages will be accounted for in `message_bus_headers_max`. This will let us reduce static memory allocation considerably since the pipeline can be long.

Sounds good, I'll start working on that and include it in this branch+PR.

sentientwaffle committed:
VSR: RingBuffer.push_assume_capacity

- In `message_bus.zig`, precheck the send queue capacity.
- Also include the `Connection` messages in the `message_bus_messages_max` calculation.
- Refactor the `message_bus_messages_max` calculation into a block scope for clarity.

sentientwaffle committed:
Config: Move {messages,headers}_max → message_pool

sentientwaffle commented:
I'm trying to figure out the deterministic message/header pool capacities.

Right now in `replica.zig`, `create_message_from_header` calls `get_header_only_message` to get a message. That (header-only) message is either sent to a client via the `MessageBus`, or to a replica via the `loopback_queue`/`MessageBus`.

But both the `loopback_queue` and `MessageBus` (`Connection.send_queue`) are used for full messages as well as header-only messages. So to guard against `Message`-starvation deadlock, the `message_bus_messages_max` capacity must pessimistically assume that the `loopback_queue` and every `Connection`'s `SendQueue` may be full messages. But likewise, `message_bus_headers_max` must assume that they are fully header-only messages. So this approach doesn't actually save memory. And it would be better off using just the full messages in this case, to reduce the maximum number of header-only messages.

Am I missing something? I know that using header-only messages specifically for `pipeline`'s `prepare_ok` messages (described in https://github.com/coilhq/tigerbeetle/issues/70) is still useful and will reduce memory overhead.

sentientwaffle committed:
VSR: Use a bitset for pipeline's prepare_ok quorum

To decrease the number of messages allocated (and thus memory usage), `pipeline` now stores only the `prepare` messages, not the `prepare_ok`s.

jorangreef commented:
> Am I missing something? I know that using header-only messages specifically for pipeline's prepare_ok messages (described in #70) is still useful and will reduce memory overhead.

100%. Great to be working out the worst-case concurrency of everything like this now.

Would it help if we had separate queues for different sizes messages on the send/receive paths? Or should we only plan for small messages on the recv path (and when sending to ourself)?

I think we should also make our header-only messages a little bigger (1KiB, just under the size of a single data packet MSS I believe?) so that we can use them for all the view change messaging and save space on our quorum counters. We just need to check that making them slightly bigger to broaden application provides an overall saving outweighing the small increase of the pipeline size.

We could then move to this kind of naming: `message_min` (1KiB) and `message_max` (1-2MiB) with MessagePool methods like `get_message_min()` and `get_message_max()`.

I'm also thinking that moving the pool allocation calculation out of config is so important. Having it in config makes it look safe to tune, like a tuning parameter for memory usage, when in fact it can cause deadlock.

sentientwaffle commented:
> Would it help if we had separate queues for different sizes messages on the send/receive paths?

I don't understand the separate queues idea — can you elaborate? 

> Or should we only plan for small messages on the recv path (and when sending to ourself)?

Setting the separate queues idea aside, I'm leaning towards yes (though I'm not sure about the "sending to ourself" part).

Even when sending a header-only message (either to other replicas or self), the queue's capacity (measured in preallocated messages) must still account for the full-size messages. So in that case allocating extra header-only messages would be redundant.

---

> I think we should also make our header-only messages a little bigger (1KiB, just under the size of a single data packet MSS I believe?) so that we can use them for all the view change messaging and save space on our quorum counters. We just need to check that making them slightly bigger to broaden application provides an overall saving outweighing the small increase of the pipeline size.

In the (current) default configuration:

- `@sizeOf(Header) == 128`
- `replicas_max    == 6`
- `pipelining_max  == clients_max == 32`
- `pipeline` contains (up to) `pipelining_max` `prepare`s, each with (up to) `replicas_max` `prepare_ok`s, for a total of `(6+1)*32 = 224` messages.
- Each of the replica's other 3 quorums has `replicas_max` messages.

So just to estimate (this isn't exact, since it ignores some metadata):
- If the `pipeline` quorums are header-only (128B), that saves `6*32*(1MiB-128)=201302016B`
- If the `pipeline` quorums *and* the other quorums are 1KiB, it saves `6*(32+3)*(1MiB-1KiB)=219985920B`.
i.e. it's 8% smaller. :+1:

Though, why does a `Prepare` actually need to hold onto the actual `prepare_ok` messages at all? (Rather than just tracking which replicas it has received a `prepare_ok` from).

---

> I'm also thinking that moving the pool allocation calculation out of config is so important. Having it in config makes it look safe to tune, like a tuning parameter for memory usage, when in fact it can cause deadlock.

:+1: I haven't pushed the change yet, but I moved those 2 constants into `message_pool.zig`, so it is apparent that they aren't configurable. In a separate PR I think we should reconsider which constants are actually configurable, and which are derived (also add constraints to the configured values, to ensure they are sane).

jorangreef commented:
Thanks for the back-of-the-envelope comparisons.

> I don't understand the separate queues idea — can you elaborate?

I was just wondering about it — if we were to keep the queues all separate per protocol message on separate circuits then we might have a shot at working everything out using the different sizes. I was more asking the question in case it might be possible. Regardless, maintaining multiple queues would increase dimensionality so let's not explore it further. I'm also not actually sure it's possible.

> Though, why does a Prepare actually need to hold onto the actual prepare_ok messages at all? (Rather than just tracking which replicas it has received a prepare_ok from).

Great question. This gets to the heart of it. Pipelining was added only recently. Before pipelining, I was focused on getting the protocol correct first before optimizing memory efficiency. Regarding quorum counting in general, I originally started out with bitsets to track quorum and then switched tack to hold onto the acks so that we could be more careful when we count the quorums (comparing checksums etc. as an extra cryptographic guarantee), since getting quorum wrong breaks everything. But again, this was before pipelining, and pipelining really amplifies the memory cost.

> Even when sending a header-only message (either to other replicas or self), the queue's capacity (measured in preallocated messages) must still account for the full-size messages.

If we switch to counting `prepare_ok` quorums using a 64-bit integer bitset (to support more than 6 replicas in future, and for nice alignment, i.e. 8-byte prepare pointer, plus 8-byte bitset count), then we can even drop the whole concept of a header-only message completely. The pipeline is currently the only place where we are able to count header-only messages for purposes of static allocation. Otherwise, as you say, we have to count all the queues as if they were containing full messages. So we could save memory and simplify at the same time.

I would be happy to use bitset quorum counters elsewhere too, where we can. Provided we are always careful to compare with the context of the prepare, cross-checking/binding the `prepare_ok` with the `prepare` as much as we  possibly can before we set the bit, and being careful for idempotency, then we would remain safe enough, and the VOPR would also be able to verify this. It could be a really nice optimization in terms of memory.

What do you think? Would you like to work on the memory optimization of the pipeline? If so, would you like to first get the static allocation count right in one PR, and then do a separate PR for the pipeline memory optimization, or would you prefer to tackle it as one?

jorangreef commented:
Also, just thinking a little ahead, if we can get to the point where we can guarantee sufficient static allocation, it could be really nice later to be able to assert on that rather than having to handle the case where a message is not available. It would help us surface any deadlock bugs in future much quicker.

sentientwaffle commented:
Yeah, I'll start on the pipeline/quorum memory optimization. Usually I'd split it into 2 PRs, but I think I'll just do the one — the VOPR memory overhead with the full static allocation and a high number of replicas+clients is a bit excessive.

> Also, just thinking a little ahead, if we can get to the point where we can guarantee sufficient static allocation, it could be really nice later to be able to assert on that rather than having to handle the case where a message is not available. It would help us surface any deadlock bugs in future much quicker.

Absolutely, that would be awesome, though I'll have to think more about how. Maybe some sort of recursing reflection over the `TypeInfo` of the replica.

jorangreef commented:
> Yeah, I'll start on the pipeline/quorum memory optimization. Usually I'd split it into 2 PRs, but I think I'll just do the one — the VOPR memory overhead with the full static allocation and a high number of replicas+clients is a bit excessive.

Awesome! Yes, the VOPR is very hungry at present. Reducing memory overhead significantly would be fantastic, also allowing us to test larger clusters.

> Absolutely, that would be awesome, though I'll have to think more about how. Maybe some sort of recursing reflection over the TypeInfo of the replica.

So I can understand better, why would we need to reflect over the TypeInfo of Replica? I was only thinking that if we know we have our static allocation worked out sufficiently, then it would just be a matter of asserting whenever `get_message()` is called — in fact, simply changing the function signature to no longer be optional and asserting within.

sentientwaffle commented:
> So I can understand better, why would we need to reflect over the TypeInfo of Replica? I was only thinking that if we know we have our static allocation worked out sufficiently, then it would just be a matter of asserting whenever `get_message()` is called — in fact, simply changing the function signature to no longer be optional and asserting within.

I was imagining a comptime assert that the replica data structure never has more than `n` references to a `Message` type, but your way is much simpler. :stuck_out_tongue_closed_eyes:

jorangreef commented:
> I was imagining a comptime assert that the replica data structure never has more than n references to a Message type, but your way is much simpler. 

That would be a fantastic assert to pull off.

sentientwaffle committed:
VSR: Store fewer quorum messages.

`start_view_change_from_other_replicas` and `nack_prepare_from_other_replicas` don't need the message objects. Save more memory by just tracking which replica has sent the message.

`do_view_change_from_all_replicas` still needs its messages.

sentientwaffle committed:
VSR: Remove header-only messages.

Since the message pool capacity is now statically determined based on the maximum possible number of required messages, header-only messages don't reduce memory.

sentientwaffle committed:
VSR: Separate replica and client message counts

A client doesn't need to allocate as many messages as a replica, so this avoids some unnecessary memory usage.

The client's `SendQueue` capacity is set to 2. It is *possible* (albeit unlikely) for it to (attempt) to exceed that — so there's no assert.

`client_request_queue_max` is broken out into a separate constant, rather than deriving it from the `messages_max_client` — the derivation is now in the other direction. This prevents the case where there are no messages available because the `RequestQueue` is full but pings (which aren't included in the request queue) are being generated and stored in the `SendQueue`.

sentientwaffle committed:
VOPR: Print number of ticks on completion

This is useful when comparing a single seed before and after a code change, to verify the performance didn't regress (e.g. extra retries).

sentientwaffle commented:
> I was only thinking that if we know we have our static allocation worked out sufficiently, then it would just be a matter of asserting whenever `get_message()` is called — in fact, simply changing the function signature to no longer be optional and asserting within.

Does it make sense to crash the replica (or client, potentially) for an error like this? If the static message pool capacity is calculated correctly, it will never happen. But if it does happen, it is usually (or maybe always since the pipeline will be using a bitset for `prepare_ok`s) recoverable with the existing `null`-handling.

One option to consider: have a VOPR/testing-only `assert` to detect the bug in testing, but still handle the error if it actually triggers.

jorangreef commented:
That's a great question to ask.

We would still prefer to crash in line with our fail-safe philosophy. That would be the right thing to do. A few riffs to motivate:

1. TigerBeetle guarantees static allocation to the operator.
2. Resource exhaustion could have an impact on variance which could lead to metastability issues, or cascading failure. This might take even longer to diagnose in production than a simple crash, leading to overly prolonged outages, cf. the recent Roblox outage report. If production does trigger a deadlock that the VOPR didn't detect, by virtue of more interesting production data, then it's easier to figure out what went wrong if we get a crash with a stack trace.
3. If we've calculated our static allocation correctly, and we do detect resource exhaustion then that could indicate a leak, or could imply buffer reuse, or something else we might not have imagined that could be dangerous. It's a logic failure where we haven't thought through all the branches, so we're in unknown territory, and it's better not to continue running.
4. It's always better to reduce dimensionality at call sites, so simplifying function signatures to not have a null case is a significant benefit, by reducing the probability that we run into bugs in other code (even simply by reducing our function bodies to under 70 lines to make control flow as a whole easier to read).
5. If we're in production and see we don't have a message, then we may ignore it, but that might not mean it's ultimately recoverable really, for example it might just be another form of deadlock that won't exit.
6. We want to run the same code as much as possible in the VOPR as we do in production. We don't want to have a bimodal profile in production that we don't have in testing.
7. It maintains our incentive to calculate the static allocation correctly. This is also one of the ideas behind Toyota's old "stop the production line at the first sign of anything wrong no matter how small", which improves quality.

sentientwaffle commented:
:+1: Great explanation, thank you! Do those criteria extend to the TB client, which runs embedded in the user's own application/service?

sentientwaffle committed:
MessagePool: Getting a message always succeeds.

The maximum number of active messages that can ever be needed by a replica/client is defined in `message_pool.zig`. As such, `get_message` can always succeed.

If (due to a bug) the limit is too low, `get_message` will panic.

See https://github.com/coilhq/tigerbeetle/pull/72 for more details.

jorangreef commented:
> 👍 Great explanation, thank you! Do those criteria extend to the TB client, which runs embedded in the user's own application/service?

It's a pleasure!

Yes, the criteria extend to the TB client, also because:

1. We want to convert hidden failure into visible failure.
2. On the client there's less opportunity for us to notify the operator that something is wrong, because it's not like we have a database log file in the data directory that we can write to.
3. If the client crashes early, it increases the chance that a developer testing the client in pre-production will see that something is up and report an issue. Whereas, trying to keep going means it's more likely to fail in production.

The golden rule for asserts, is simply always to enumerate and assert everything exhaustively, all pre-conditions, all post-conditions, all interim steps, and to stretch our understanding completely across the positive space and beyond into the negative space. Crashes are immensely valuable, and a way to learn and iteratively correct our understanding. We then reduce the likelihood that there will be a string or chain of bugs (really just semantic gaps in our understanding) that do damage in production.

We're always downgrading correctness bugs into liveness bugs. If we know our understanding is not correct, then the best thing to do is fail-fast, fail-safe.

This gives the operator confidence that TigerBeetle will always run correctly, or shutdown safely if the slightest thing is wrong. Assertions are also counter-intuitive, the more of them there are, the less likely we are to crash, because assertions are a forcing function to check and also document our understanding in the code.

Note that I'm thinking here of programming/logic errors (where what we expect doesn't happen, or where what we don't expect does happen — we have to assert both), not operating errors. The only safe way to handle logic errors is to stop executing logic, because once we become aware of an unsafe logic path, to do further logic is to compound the error (the result is completely undefined).

Whereas, operating errors are where we expect the machine/storage/network to fail. We do need to recover from operating errors without crashing. For example, that's why we go to such lengths with our storage fault model and recovery from corruption. This is where TigerBeetle should provide high availability without crashing, whereas many systems would simply crash.

sentientwaffle committed:
VSR: Add a couple asserts

jorangreef commented on src/message_pool.zig:
```diff
@@ -16,15 +16,48 @@ comptime {
+    const connection_messages_max = config.connection_send_queue_max_replica + 1;
+    const message_bus_messages_max = config.connections_max * connection_messages_max;
+
+    break :messages_max pipelining_messages_max + quorum_messages_max +
```
Let's either sum the variables in the same order that we calculate them to make it easier to compare (e.g. `client_table_messages_max + journal_messages_max` etc.), or else let's get rid of the variables entirely and simply add to `var sum = 0` multiple times as we calculate, once per line, with a comment to explain each line's calculation, since most variables are used only once.

jorangreef commented on src/config.zig:
```diff
@@ -102,8 +96,15 @@ pub const pipelining_max = clients_max;
-/// The maximum number of outgoing messages that may be queued on a connection.
-pub const connection_send_queue_max = pipelining_max;
+/// The maximum number of outgoing messages that may be queued on a replica connection.
+pub const connection_send_queue_max_replica = pipelining_max;
```
I realize that `connection_send_queue_max` was originally set by me to be too high.

The MessageBus connection send queue is literally our user space TCP send buffer of stuff just waiting to cross to kernel space through io_uring. It's a small short term bursty queue, representing how long it takes an io_uring `send()` syscall to do the memcopy to the kernel's TCP send buffer, which we set via `config.tcp_snd_buf`, all of which is with respect to only a single TCP connection.

It can therefore be set smaller than `pipelining_max`, since the pipeline is long term, representing the entire lifetime of inflight prepare messages (either about to be copied to the kernel, or about to be sent by the kernel over a TCP connection, or sent and awaiting ack or retransmit).

The difference between send queue and pipeline, is the rate of sending or "memcopy bandwidth to the kernel" for send queue vs "network bandwidth-delay product" for pipeline.

A better value for `send_queue_max` may be the number of requests from clients that the primary replica expects to receive all in the very same instant that it then needs to push to the send queue of the next replica, plus space for a control plane message, so perhaps `max(min(clients_max, 4), 2)`?

It shouldn't be set higher than the number of messages that can fit into `tcp_snd_buf` or we would be queueing beyond what the network link can send.

jorangreef replied:
It would also be great while we are thinking about this, to assert that `tcp_snd_buf >= message_size and tcp_snd_buf <= message_size * 2`, since setting TCP buffers too large can cause compaction latency spikes in the kernel when it does packet bookkeeping (Cloudflare ran into this—we have a link to the incident in `config.zig`).

sentientwaffle replied:
> A better value for `send_queue_max` may be the number of requests from clients that the primary replica expects to receive all in the very same instant that it then needs to push to the send queue of the next replica, plus space for a control plane message, so perhaps `max(min(clients_max, 4), 2)`?

I don't understand the `max(min(..)..)` recommended value; can you elaborate? Based on your description, wouldn't it just be something like `clients_max + 1`?

---

Also, why not set the `tcp_snd_buf` to `connection_send_queue_max * message_size_max`?

jorangreef replied:
> Also, why not set the tcp_snd_buf to connection_send_queue_max * message_size_max?

The reason for not setting `tcp_snd_buf` too far beyond 4 MiB is an interesting production incident that happened to Cloudflare (and the kernel by default also typically sets the send buffer to at most 4 MiB per connection):

```
/// At the same time, be careful going beyond 4 MiB as the kernel may merge many small TCP packets,
/// causing considerable latency spikes for large buffer sizes:
/// https://blog.cloudflare.com/the-story-of-one-latency-spike/
```

> I don't understand the max(min(..)..) recommended value; can you elaborate? Based on your description, wouldn't it just be something like clients_max + 1?

`clients_max + 1` would be if all clients simultaneously send a request over the network that arrive all in a single flush of our io_uring completion queue, faster than our memcopies to the kernel are able to empty the send_queue from user space to kernel space. This is an edge scenario, a skewed distribution. In practice, there would really be an ms or two between client request arrivals (i.e. multiple flushes of the completion queue), and our copies to kernel space when calling `send()` should be able to empty the send queue much faster than that rate. So the `max(min())` above (which would work out to around 2-4 messages in the send queue) should be sufficient.

It comes down to getting a good estimate on how many `recv()` syscall results we expect to flush from the completion queue in a single call to `io_uring_enter()`. Do we expect this to be between 25-50% of `clients_max` in the 99% case? Note also that io_uring will wake us the instant that we receive a single `recv()` syscall, and this already gives us a chance to do the corresponding memcopy through the send queue back to kernel space when replicating the prepare, while we go back to getting the next `recv()`. In other words, the send queue likely has a faster turnaround time than the recv queue, since the former is operating at memory bandwidth, whereas the latter is operating at network bandwidth. We could also plug this into Little's Law to get an idea of how the queue sizes should compare with respect to each other.

Clients will retry if the edge scenario happens, so there's no static allocation issue, only a latency spike issue, but it's very rare. We need to balance it against the memory cost of `clients_max` messages for all `replicas_max` connections, which is significant.

sentientwaffle replied:
Thank you, that makes sense.

I did read the cloudflare article, but that's not quite what I meant in my question. 

As I understand it, the `send_queue_max` is also the maximum number of messages that will go in the tcp send buffer. And since messages are (at most) 1M, that comes out to 4M when the `send_queue_max` is at its (new) maximum value of `4`. Isn't `send_queue_max * message_size_max` precisely the right value for `tcp_snd_buf`? Any less and the send queue might not be able to get all its data into the kernel. And more and it runs into the issue described in that blog post.

jorangreef replied:
Thanks for clarifying the motivation for setting `tcp_snd_buf ` to `send_queue_max * message_size_max`.

Let's do that, now that we have a tighter limit on the send queue. It does seem like we're approaching the new maximum of `4` from both directions this way. We could also double-check this against the typical NIC bandwidth we want to support. For example, how much snd buffer do we need to saturate a 1 Gbps or 10 Gbps NIC?

jorangreef commented on src/message_bus.zig:
```diff
@@ -586,15 +588,13 @@ fn MessageBusImpl(comptime process_type: ProcessType) type {
+                    });
+                    return;
+                }
+                connection.send_queue.push_assume_capacity(message.ref());
```
Nice!

jorangreef replied:
Perhaps we should use `err` as the log level here since dropping messages might be something we want to give more urgency to if/when it happens?

jorangreef commented on src/message_bus.zig:
```diff
@@ -738,14 +738,7 @@ fn MessageBusImpl(comptime process_type: ProcessType) type {
-                    connection.terminate(bus, .shutdown);
-                    return null;
-                };
+                const message = bus.get_message();
```
We need to `+1` to our recv message count, since this is in addition to the single `recv_message` that the connection already has in our static allocation calculation.

jorangreef replied:
There may be a few other places like this, where we call `get_message()` and temporarily burst over static allocation.

sentientwaffle replied:
:+1: good catch. I did account for one of those in the client, but hadn't considered it more generally.

I think it's enough to just add `+ 1` to the `messages_max_replica` (the client already has it) to handle _all_ the bursts. That is, bursts can never overlap, because the burst message will either be stored in the respective queue/pipeline/etc, or dropped if the respective queue/... is full.

jorangreef replied:
Good, and it's also cheaper and more accurate to add the `+ 1` to the total, rather than for each connection, since it only happens one-by-one in the control plane.

sentientwaffle replied:
Yes, exactly!

jorangreef commented on src/message_pool.zig:
```diff
@@ -16,15 +16,48 @@ comptime {
+    const connection_messages_max = config.connection_send_queue_max_client + 1;
+    const message_bus_messages_max = config.replicas_max * connection_messages_max;
+    // +1 to account for creating a ping when the send/request queues are already full.
+    break :messages_max 1 + message_bus_messages_max + config.client_request_queue_max;
```
Are we perhaps double-counting one of the client's message bus messages when we add these two together, since the first request in the client request queue is shared with the message bus?

I'm not 100% sure, so please double-check me, but it seems that a message can only be referenced in the message bus if it is either a control plane message (e.g. ping), or an inflight request (1 at a time max) or a reply, and the inflight request is already accounted for in the client request queue.

jorangreef replied:
For example, if there are 6 replicas, and we expect 1 inflight request plus 1 ping plus 1 reply per replica, then we might think this could be 12 messages max (not 18, since the 1 inflight request is already in the request queue). If our client request queue is only 10 messages max, and if 1 of these is already counted as the inflight request, then we might expect the static allocation count to be `10 + 6*(1+1) = 22` rather than 28?

sentientwaffle replied:
I did consider this case. But (and I admit this is an awfully obscure edge case) is there anything preventing both positions of the connection send queue from being (separate) pings?

jorangreef replied:
Great question.

Perhaps we shouldn't try to get a ping message if the send queue is full? This might also be something to consider... not attempting to get a message if we see we can't ref it somewhere, because otherwise we might overflow our static allocation again.

jorangreef replied:
Although I think that's already covered by our `always add 1 to the total` principle to cover temporary overflow in the control plane.

sentientwaffle replied:
Yeah, I think "always add 1" means that it's safe (i.e. won't crash).

> Perhaps we shouldn't try to get a ping message if the send queue is full?

If I remember right, the client always pings all replicas at once (using a single many-reference ping message) so I think its probably simpler as-is, even though it can (rarely) create a ping that is immediately discarded by the message bus if every connection's send queue is full. It isn't much extra work, and a single message overhead in the static allocation is pretty minor (and is simpler than checking every send queue prior to the ping).

Despite the shared ping message, the client still needs to account for `replica_count` ping messages since they can send at different speeds (again, very much an obscure edge case).

jorangreef commented on src/message_pool.zig:
```diff
@@ -16,15 +16,48 @@ comptime {
+};
+
+comptime {
+    // These conditions is necessary (but not sufficient) to prevent deadlocks.
```
```suggestion
    // These conditions are necessary (but not sufficient) to prevent deadlocks.
```

Great checks to have. We definitely would be surprised if `messages_max` were less, so it's essential to assert that.

jorangreef commented on src/message_pool.zig:
```diff
@@ -16,15 +16,48 @@ comptime {
         header: *Header,
-        /// Unless this Message is header only, this buffer is in aligned to config.sector_size
-        /// and casting to that alignment in order to perform Direct I/O is safe.
+        /// This buffer is in aligned to config.sector_size and casting to that alignment in order
```
```suggestion
        /// This buffer is aligned to config.sector_size and casting to that alignment in order
```

jorangreef commented on src/message_pool.zig:
```diff
@@ -74,42 +103,16 @@ pub const MessagePool = struct {
+    /// Get an unused message with a buffer of config.message_size_max.
+    /// The returned message has exactly one reference.
+    pub fn get_message(pool: *MessagePool) *Message {
+        const ret = pool.free_list.?;
```
```suggestion
        const message = pool.free_list.?;
```

jorangreef commented on src/ring_buffer.zig:
```diff
@@ -76,6 +76,13 @@ pub fn RingBuffer(comptime T: type, comptime size: usize) type {
+        /// Add an element to a RingBuffer, and assert that the capacity is sufficient.
+        pub fn push_assume_capacity(self: *Self, item: T) void {
+            self.push(item) catch |err| switch (err) {
+                error.NoSpaceLeft => unreachable,
```
Good to see that the error set is explicitly switched on all errors, rather than simply doing `catch unreachable`.

jorangreef commented on src/vsr/client.zig:
```diff
@@ -66,8 +67,7 @@ pub fn Client(comptime StateMachine: type, comptime MessageBus: type) type {
         /// A client is allowed at most one inflight request at a time at the protocol layer.
         /// We therefore queue any further concurrent requests made by the application layer.
-        /// We must leave one message free to receive with.
-        request_queue: RingBuffer(Request, config.message_bus_messages_max - 1) = .{},
+        request_queue: RingBuffer(Request, config.client_request_queue_max) = .{},
```
Nice to make this constant explicit for the sole purpose of the request queue.

jorangreef commented on src/vsr/client.zig:
```diff
@@ -402,8 +402,7 @@ pub fn Client(comptime StateMachine: type, comptime MessageBus: type) type {
 
-            var message = self.message_bus.get_message() orelse
-                @panic("register: no message available to register a session with the cluster");
+            var message = self.message_bus.get_message();
```
Can we do `const message = ` here and also grep for other places in the source where we used to do this?

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -52,7 +52,7 @@ const Prepare = struct {
 
     /// Unique prepare_ok messages for the same view, op number and checksum from ALL replicas.
-    ok_from_all_replicas: QuorumMessages = QuorumMessagesNull,
+    ok_from_all_replicas: QuorumCounter = QuorumCounterNull,
```
`QuorumCounter` is a better name for our quorum counter 😄

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -735,7 +729,7 @@ pub fn Replica(
             const threshold = self.quorum_replication;
 
-            const count = self.add_message_and_receive_quorum_exactly_once(
+            const count = self.tally_message_and_receive_quorum_exactly_once(
```
```suggestion
            const count = self.count_message_and_receive_quorum_exactly_once(
```

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -882,28 +876,22 @@ pub fn Replica(
-                }
+                const available = self.message_bus.get_message();
+                self.message_bus.unref(available);
             }
```
We can drop this entire branch now that we guarantee that we can always send a `do_view_change` message to our loopback queue.

We also need to drop it because if we have already sent the `do_view_change` to ourself then this could overflow our static allocation.

Let's check that we don't have a similar bug in `send_do_view_change_message` if we're sending to ourself and we've already sent it and then we try to send again.

sentientwaffle replied:
Good catch! I found several functions that still (incorrectly) return a `?*Message` instead of a `*Message`.

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -1484,8 +1462,12 @@ pub fn Replica(
             var waiting_len: usize = 0;
-            for (prepare.ok_from_all_replicas[0..self.replica_count]) |received, replica| {
-                if (received == null and replica != self.replica) {
+            var ok_from_all_replicas_iter = prepare.ok_from_all_replicas.iterator(.{
```
```suggestion
            var ok_from_all_replicas_iterator = prepare.ok_from_all_replicas.iterator(.{
```

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -1484,8 +1462,12 @@ pub fn Replica(
+                .kind = .unset,
+            });
+            while (ok_from_all_replicas_iter.next()) |replica| {
+                if (replica == self.replica_count) break;
```
Let's add a comment to explain why we must do this when iterating the quorum counter.

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -1670,6 +1633,75 @@ pub fn Replica(
+            }
+
+            messages.set(message.header.replica);
+            assert(messages.count() <= self.replica_count);
```
Good assert to have. We could also test the std lib bit set by asserting `messages.isSet` after doing the set. It's in the control plane here, no trouble for us, and we would be surprised if it wasn't set after having just set it.

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -1670,6 +1633,75 @@ pub fn Replica(
+                return null;
+            }
+
+            messages.set(message.header.replica);
```
```suggestion
            // Record the first receipt of this message:
            messages.set(message.header.replica);
```

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -1670,6 +1633,75 @@ pub fn Replica(
 
+        fn tally_message_and_receive_quorum_exactly_once(
+            self: *Self,
+            //prepare: *Prepare,
```
What are your thoughts on taking the `prepare` here as an arg? Was it to double-check the `prepare_ok.context` against the `prepare.checksum`?

sentientwaffle replied:
Oops, that is old code. Yes, it originally was to double-check the prepare. But then I changed `start_view_change_from_other_replicas` and `nack_prepare_from_other_replicas` from `QuorumMessages` to `QuorumCounter`, so the function has to be more general.

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -1670,6 +1633,75 @@ pub fn Replica(
+        fn tally_message_and_receive_quorum_exactly_once(
+            self: *Self,
+            //prepare: *Prepare,
+            messages: *QuorumCounter,
```
```suggestion
            counter: *QuorumCounter,
```

jorangreef replied:
I think `counter` maps nicely to the `messages` arg that we use in the corresponding `add_message...` function, and also in the same way that we have `QuorumCounter` and `QuorumMessages`.

Perhaps we could rename our `add_message...` method for `do_view_change` messages to then be `reference_message_and_receive_quorum_exactly_once` so that the two methods are more symmetrical according to their function (one refs, the other counts)?

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -1833,10 +1865,7 @@ pub fn Replica(
             }
 
             // TODO We can optimize this to commit into the client table reply if it exists.
```
We need to do this now, using the existing reply if it exists in the client table.

Otherwise we would temporarily overflow our static allocation count for the client table, by having a message reference in there already and then acquiring a new one now.

Alternatively, if reusing the existing reply message is too complicated (perhaps because it might be reffed by the send queue and so in the progress of being copied to the kernel), then we should rather +1 to our total client table count, provided we know that we will immediately unref the old reply being replaced.

sentientwaffle replied:
> Alternatively, if reusing the existing reply message is too complicated (perhaps because it might be reffed by the send queue and so in the progress of being copied to the kernel), then we should rather +1 to our total client table count, provided we know that we will immediately unref the old reply being replaced.

We could test that `references == 1` before modifying, and when `references != 1` just replace it like usual. But I'm leaning towards just increasing the message count +1 anyway, to avoid that (rather complicated) case.

The existing code can still be improved though. Since `get_message` is infallible, it can be moved into `commit_op` rather than making the caller get the message and pass it in.

jorangreef replied:
> But I'm leaning towards just increasing the message count +1 anyway, to avoid that (rather complicated) case.

Agreed.

> The existing code can still be improved though. Since get_message is infallible, it can be moved into commit_op rather than making the caller get the message and pass it in.

That will be great, reducing the dimensionality of the `get_message` return type is simplifying the code nicely already. We had several bugs found by the VOPR around messages not being available with respect to committing or with respect to retrying a send after a timeout. Working out the static allocation like we are doing now requires care and thought, but the alternative meant more complexity and bugs at the same time. Furthermore, our understanding around resource usage is growing because the assertions are pushing us to explore the boundary between positive/negative space.

Definitely looking forward to getting a pure crash in the VOPR instead of a stuck bug if we have any further static allocation issues in future. 😄

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -1937,27 +1966,19 @@ pub fn Replica(
-                    .prepare_ok,
-                    prepare.message.header.checksum,
-                );
+                const count = prepare.ok_from_all_replicas.count();
```
So clean!

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -1937,27 +1966,19 @@ pub fn Replica(
-                    prepare.message.header.checksum,
-                );
+                const count = prepare.ok_from_all_replicas.count();
                 assert(count >= self.quorum_replication);
```
Let's assert what we expect to be the upper bound for `count` here too.

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -3457,18 +3463,28 @@ pub fn Replica(
         }
 
+        fn reset_quorum_counter(self: *Self, messages: *QuorumCounter) void {
+            var messages_iter = messages.iterator(.{});
```
```suggestion
            var counter_iterator = counter.iterator(.{});
```

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -3457,18 +3463,28 @@ pub fn Replica(
+            }
+
+            messages.setIntersection(QuorumCounterNull);
+            assert(messages.count() == 0);
```
Great to see as this assert is critical. If the quorum is somehow not reset fully then we would have a correctness bug leading to data loss as quorums become overinflated without any real backing from real acks.

jorangreef replied:
Let's also add a loop up to `replica_count` asserting that all replica bits are unset, since this is so critical.

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -3573,11 +3589,7 @@ pub fn Replica(
-                0,
-            );
+            const count_start_view_change = self.start_view_change_from_other_replicas.count();
             assert(count_start_view_change >= self.quorum_view_change - 1);
```
Let's also assert our expected upper bound for the count.

jorangreef reviewed (changes requested):
Looking good. This will be great when it lands.

What are the "before and after" memory allocation savings that we might see from this, now that we're using bitset counters instead of taking message references in the pipeline?

sentientwaffle commented:
I'm still going through your review suggestions, but re:

> What are the "before and after" memory allocation savings that we might see from this, now that we're using bitset counters instead of taking message references in the pipeline?

Bitset counters save `(pipelining_max:32 + start_view_change:1 + nack_prepare:1) * (replicas_max:6) * message_size_max:1M = 204M`.

Which is helpful, but not huge — the majority of memory saved (in the simulator) is from setting the client's `messages_max` separately from the replica's. Especially because the new (computed) value of `messages_max_replica` is so high.

sentientwaffle committed:
VSR: tidy, remove old optionals (code review)

- Set a much smaller `connection_send_queue_max_replica` using the value recommended by Joran.
- Split `tcp_sndbuf` into 2 configs (for replica and client).
- Derive the `tcp_sndbuf` value from the send queue size.
- Add more assertions.
- Eliminate unneeded code due to `get_message()` being infallible.
- Count `commit_op`'s reply.

sentientwaffle commented:
More details:

- Before PR:
  - Replica: 152 messages + 2432 header-only messages.
  - Client: same as replica
- After PR:
  - Replica: 279 messages + 0 header-only messages.
  - Client: 51 messages + 0 header-only messages.

Header-only messages are a rounding error for memory usage, but removing them simplifies the code. Also, due to how static message limits are computed, they aren't useful.

So this change increases the replica's memory usage (to fix the deadlock), and decreases the client's memory usage. Also, if (due to a mistake in the static allocation) the replica runs out of messages: it will just crash (which can be recovered from automatically) rather than deadlocking (which requires manual administrator intervention).

jorangreef commented on src/message_pool.zig:
```diff
@@ -16,15 +16,53 @@ comptime {
+pub const messages_max_replica = messages_max: {
+    var sum: usize = 0;
+
+    sum += config.io_depth_read + config.io_depth_write; // journal I/O
```
```suggestion
    sum += config.io_depth_read + config.io_depth_write; // Journal I/O
```

jorangreef commented on src/message_pool.zig:
```diff
@@ -16,15 +16,53 @@ comptime {
+    var sum: usize = 0;
+
+    sum += config.io_depth_read + config.io_depth_write; // journal I/O
+    sum += config.clients_max; // Replica.client_table
+    sum += 1; // Replica.loopback_queue
+    sum += config.pipelining_max; // Replica.pipeline
+    sum += config.replicas_max; // Replica.do_view_change_from_all_replicas quorum (all others are bitsets)
+    sum += config.connections_max; // Connection.recv_message
+    sum += config.connections_max * config.connection_send_queue_max_replica; // Connection.send_queue
+    sum += 1; // Handle bursts (e.g. Connection.parse_message)
+    // Handle Replica.commit_op's reply.
+    // (This is separate from the burst +1 because they may occur concurrently).
+    sum += 1;
+
+    break :messages_max sum;
+};
+
+/// The number of full-sized messages allocated at initialization by the client message pool.
+pub const messages_max_client = messages_max: {
+    var sum: usize = 0;
+
+    sum += config.replicas_max; // Connection.recv_message
+    sum += config.replicas_max * config.connection_send_queue_max_client; // Connection.send_queue
+    sum += config.client_request_queue_max; // Client.request_queue
+    // Handle bursts (e.g. Connection.parse_message, or sending a ping when the send queue is full).
+    sum += 1;
+
+    break :messages_max sum;
+};
```
Very nice.

jorangreef commented on src/config.zig:
```diff
@@ -218,3 +223,9 @@ pub const clock_synchronization_window_min_ms = 2000;
+    // Avoid latency issues from a too-large sndbuf.
+    assert(tcp_sndbuf_replica <= 4 * 1024 * 1024);
+    assert(tcp_sndbuf_client <= 4 * 1024 * 1024);
+}
```
Great to have these here.

jorangreef commented on src/test/state_checker.zig:
```diff
@@ -8,12 +8,13 @@ const Cluster = @import("cluster.zig").Cluster;
 const RingBuffer = @import("../ring_buffer.zig").RingBuffer;
 
-const RequestQueue = RingBuffer(u128, config.message_bus_messages_max - 1);
+const RequestQueue = RingBuffer(u128, message_pool.messages_max_client - 1);
```
```suggestion
const RequestQueue = RingBuffer(u128, config.client_request_queue_max);
```

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -1484,8 +1449,17 @@ pub fn Replica(
+            });
+            while (ok_from_all_replicas_iterator.next()) |replica| {
+                // Ensure we don't wait for replicas that don't exist.
+                if (replica == self.replica_count) {
```
Are we guaranteed to hit this exit condition always?

The comment is good. It would be good to further explain a little more why the captured `replica` might go beyond `replica_count`.

sentientwaffle replied:
> Are we guaranteed to hit this exit condition always?

No, only when `self.replica_count < config.replicas_max` (since the bitset has `config.replicas_max` bits, and we are iterating over _unset_ bits, but only want to count the ones below `self.replica_count`). The assert in this block hinted at it, but I'll add more of an explanation in the comment.

sentientwaffle replied:
https://github.com/coilhq/tigerbeetle/pull/72/commits/0b383364f0d0cd4a2b62543eb36004b024b15b84

jorangreef replied:
Awesome, brilliant while loop termination assertion as well.

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -1484,8 +1449,17 @@ pub fn Replica(
+                if (replica == self.replica_count) {
+                    assert(self.replica_count < config.replicas_max);
+                    break;
+                }
```
```suggestion
                }
                assert(replica < self.replica_count);
```

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -1849,7 +1867,7 @@ pub fn Replica(
             self.commit_ops_read();
         }
 
-        fn commit_op(self: *Self, prepare: *const Message, reply: *Message) void {
```
Great to be able to simplify this interface.

jorangreef reviewed (approved):
LATM!

I left a few small suggestions and one question around the `ok_from_all_replicas` iterator. Ready to merge when those are in!

Awesome work @sentientwaffle this is very nice.

sentientwaffle committed:
VSR: Explain bitset iteration exit condition
